275313-rework-mixer-UI

blackears opened the pull request:
Changes to the mixer to make it easier to use and improve the appearance.

blackears commented:
A preview of the changes:
https://youtu.be/X3a9N3ZISJU

Some things of note: as part of my changes, I promoted Channel from being a struct to being a class so that I could encapsulate the fields.  This caused me to have to alter a dozen or so files not otherwise associated with these mixer improvements.

Setting the port and channel numbers is not fully implemented.  I found the code confusing - it seems to be checking if any other channel was currently using the same port/channel combination and assigning new values if there was a match.  I wasn't sure if this was correct behavior or not.  The old code is in MixerDetails::midiChannelChanged commented out.  I could use someone taking a look at this and recommending a good way to go about implementing it.

IsaacWeiss commented:
Comments based on watching the video:

- Values for volume and pan should be converted to be out of 100 instead of 128. Base-2 numbers are familiar to us as programmers, but to the typical user, middle pan being 64 is a mystery.
- The oft-repeated request to allow various channels belonging to a single instrument to be collapsed has almost always specified that they should be hidden by default when unused but shown by default once in use. (Side note—how are you accommodating mid-staff instrument changes, which add new channels?)
- I don't like the difference in behavior of the "Violin" channel when expanded (visually the same as the one that is visible when collapsed, despite the controls having different effects)—perhaps the one master channel should be "Violin-all" and remain visible in addition to the three when expanded.

blackears commented:
The current version presumes one instrument per part, but I can show those channels too (an oversight on my part).

One way to address points 2 and 3 is to replace the expand/collapse button with three radio buttons that switch between one track per part/one track per instrument/one track per channel.  One track per part is the reduced view and one track per channel shows everything.  

You could also add a checkbox for Hide Unusued Channels, although it's not clear to me how I would determine if a particular channel was not being used.  I think a lot of users just want the extra clutter removed and would be satisfied with being able to switch to a one track per part view.

Another option would be to add a button to each part track that would let you expand/collapse optional per channel tracks that would appear next to them.

blackears commented:
I've reworked things so that the expand button is now on the tracks that represent parts.

anatoly-os commented:
@blackears I would think about reorganizing volume mute/solo bars to horizontal layout. As you can see, Instrument names don't fit small rectangle at the bottom. There is a bit of inspiration from moder design solution: 
![buoxtqcu](https://user-images.githubusercontent.com/18065034/44837428-e337d300-ac39-11e8-8d3c-c4d6ec4fdedd.jpeg)

blackears commented:
I've made a few more change, including trying the horizontal layout:
https://youtu.be/8NN6zgyMMpw

I feel the horizontal layout has an issues.  If you scroll vertically through the controls, it's easy to accidentally change the volume and pan sliders which also respond to the mouse wheel event.  I'd also like to hear feedback on further ways to organize the display of information.

anatoly-os commented:
@blackears this is great! I like it much more than vertical controls.
I have few suggestions but they are minor, like add special mark in the volume bar at 100% level and automatically stick the button to 100% when user moves slider around (i.e. 98-102%).

Scrolling window and changing values at the same time is common problem in MuseScore interfaces (not only in MuseScore). So, for now it is not major bug.

anatoly-os commented:
After a number of discussions, the vertical sliders won. So, let's keep natural vertical sliders and use short instrument names in the instr name field.
I like new mixer so much, so let's finalize this prototype to allow users to test it in MuseScore.

blackears commented:
I've switched the orientation back round to vertical and threw in a couple of bug fixes too.  However, Travis is getting an error on this.  Not sure why, since I've squashed everything and it still builds fine on my machine.

lasconic commented:
I restarted travis. Let's see.

anatoly-os commented:
@blackears I reviewed new mixer UI. It looks much better than the current one. It looks amazing actually! Thank you for your work. It is a new step in MuseScore UI.
Btw, despite on the first time experience, it has a lot of small things to improve. My suggestions are:
1. Add tooltips for each button
2. Put right panel to the top above all the instruments controls as it is on a real mixer panel
3. Replace reverb, chorus controls, with [0-100] range values. Pan should be in range from -100 to 100. Volume should be in percents or db level so in 0-150 range or as it is done in synth. Other opinions are appreciated.
4. Replace instrument names with short instrument names at the bottom
5. Changing volume and other values of some instruments should activate and highlight this instrument in right panel
6. "M" should be center aligned as it is done for "S"
7. Add visual line for 100% volume of the instrument
8. Add master volume here (questionable)
9. "S" works different from "Solo" meaning. It works like "if one of the instruments is "S", play only instruments with "S". Combination of "S" and "M" also works interesting. So, new names are needed for checkboxes or previous logic should be applied.
10. What is the purpose of the arrow at thr top left corners of each instrument line?
11. If user changes the color of an instrument, I would change also the scroll color and pan widget color as well.

blackears commented:
@anatoly-os Glad you like the changes.  A few thoughs on the points you made:

3) This is not as easy as it might first seem.  These values are represented as chars internally and as char values in the Channel structure.  The MIDI spec expects these values to be in the range 0-127.  Mapping this onto a 1 - 100 range is going to cause round off errors and will make it hard for multiple sliders to synchronize to the same value.  It would be better to represent these values in Channel as a double and map them to the 0-127 range only when sending MIDI OUT events - however, there may be side effects for existing saved files and other things that depend on the current 0-127 range.

4) The instrument names are a complicated web.  First you have the Part (which has name that the user can set) which itself will contain one or more Instruments (should the user add more instruments to a given staff) which in turn will have one or more Channels (some insruments such as Violins will have extra channels for tremolo and pizzicato).  There's also the Patch name that describes the actual sound that the track is playing.  A combination of all of these forms the name of the instrument.  Right now I'm displaying the part name for Part tracks and the Insrument-Channel name for Channel tracks.  I'm not sure what you mean by 'short instrument names'.

6) The M is center aligned when I run it on my computer.  It's in a grid layout that has horizontalSpacers that should keep the center column aligned.  Could you post a picture of what you're seeing?

8) Not sure what you are referring to.

9) The solo works the way you are describing when I test it.  What are you seeing?

10) The arrow button expands a part so that you can see the instruments and channels within it.  So if you have one part that has an insrument change and instuments with multiple channels (such as a violin part that changes to an organ part way through) and you press the arrow button, it will expand to let you adjust these sub parts individually.

anatoly-os commented:
@blackears thank you for instant feedback and further improvements!
4. There are short names in the instruments.xml. I'm not sure now how they are read from code, but I believe what you see on all systems, but the first to the left of the staves are short instrument names.
6. See attached picture 
8. Synthesizer has Master volume indicator to the right which changes the volume of whole score. Real mixers have master volume "widget", so we can add it to the mixer. But again, I'm not sure it is correct. @lasconic what do you think about putting master volume control to the mixer UI?
9. Yes, but Solo in old Mixer behaves different. And old behavior means exactly "Solo". I'm not sure current behavior is worse, I actually like it. But we need to discuss it and if we keep it like this, the name should be changed.

Overall progress is amazing. The improvements make Mixer much better and almost ready to use. More comments about current state:
1. Top block:
1.1 We need to fix the size of horizontal volume bar in the top block
1.2 All controls including Text labels in the top block should be disabled if nothing selected (looks like they don't, but it could be just specific Windows visual feedback, see attached picture)
1.3 Imposition of top block is a bit incorrect since it breaks on resizing. There are should be fixed vertical size of the block. It should be minimal required vertical size. Horizontal size should be revised. There are two options: we can fix horizontal size and don't change it on resize or we can fix size of all elements, but pan/volume sliders.
1.4 Some settings for chorus/reverb sliders are different. See attached picture, fractions for chorus and reverb are different.
1.5 Elements of the left part of the top block should do have fixed distance between them. So, there will be empty space between the Drumkit checkbox and Mixer controls.
2. Don't move controls (Mute and Solo) up when we expand part controls (see attached picture)
3. I don't get the meaning of volume range values. 90 is default. What does 90 mean? Is it 90% from max available volume?
4. We cannot drop selection in the Mixer once we select one of the instruments. Let's select the first one when we open Mixer so we will never get not selected state. Be careful, Mixer may exist and be opened when there are no instruments, so don't forget to 
process this case.
5. The most complex question is Dark mode. Current colors suite to Dark Mode, but look foreign if we use Light mode.
![mute_shift](https://user-images.githubusercontent.com/18065034/45824087-74591180-bcef-11e8-8b30-fbf8445d1217.PNG)

blackears commented:
@anatoly-os 

9. The system I'm using is one I've seen implemented in other music programs, and they use the term 'solo' to mean 'play only instruments marked solo, excluding any that are muted'.  Having multiple instruments marked 'solo' is a little ambiguous anyhow.  I think leaving this named 'solo' would be fine.

I was looking through the MusicXML spec to fix something and saw that they define the pan range to be from -180 to 180, instead of -100 to 100, which is what I've changed it to.  It might make sense to use the -180 to 180 range, as this coresponds to the degrees of a circle.

I'm having a lot of trouble getting Travis to adopt to the new file format.  I won't be able to get much new work done until I can get that settled.

lasconic commented on awl/knob.cpp:
```diff
@@ -20,6 +20,8 @@
 
 #include "knob.h"
 
+#include <QPen>
```
Should not be necessary since it's in the precompiled header
